bus/bus.go

```diff
@@ -646,36 +646,74 @@ func (b *bus) searchObjectsHandlerGET(jc jape.Context) {
 	jc.Encode(keys)
 }
 
-func (b *bus) objectsKeyHandlerGET(jc jape.Context) {
-	ctx := jc.Request.Context()
-	if strings.HasSuffix(jc.PathParam("key"), "/") {
-		offset := 0
-		limit := -1
-		prefix := ""
-		if jc.DecodeForm("offset", &offset) != nil || jc.DecodeForm("limit", &limit) != nil || jc.DecodeForm("prefix", &prefix) != nil {
-			return
-		}
-		keys, err := b.ms.Objects(ctx, jc.PathParam("key"), prefix, offset, limit)
-		if jc.Check("couldn't list objects", err) == nil {
-			jc.Encode(api.ObjectsResponse{Entries: keys})
-		}
+func (b *bus) objectsHandlerGET(jc jape.Context) {
+	path := jc.PathParam("path")
+	if strings.HasSuffix(path, "/") {
+		b.objectEntriesHandlerGET(jc, path)
 		return
 	}
-	o, err := b.ms.Object(ctx, jc.PathParam("key"))
-	if jc.Check("couldn't load object", err) == nil {
-		jc.Encode(api.ObjectsResponse{Object: &o})
+
+	o, err := b.ms.Object(jc.Request.Context(), path)
+	if errors.Is(err, api.ErrObjectNotFound) {
+		jc.Error(err, http.StatusNotFound)
+		return
+	}
+	if jc.Check("couldn't load object", err) != nil {
+		return
 	}
+	jc.Encode(api.ObjectsResponse{Object: &o})
 }
 
-func (b *bus) objectsKeyHandlerPUT(jc jape.Context) {
+func (b *bus) objectEntriesHandlerGET(jc jape.Context, path string) {
+	var offset int
+	if jc.DecodeForm("offset", &offset) != nil {
+		return
+	}
+	limit := -1
+	if jc.DecodeForm("limit", &limit) != nil {
+		return
+	}
+	var prefix string
+	if jc.DecodeForm("prefix", &prefix) != nil {
+		return
+	}
+
+	// look for object entries
+	entries, err := b.ms.ObjectEntries(jc.Request.Context(), path, prefix, offset, limit)
+	if jc.Check("couldn't list object entries", err) != nil {
+		return
+	}
+
+	// return early if we have found entries or if a prefix was specified
+	if len(entries) > 0 || prefix != "" {
+		jc.Encode(api.ObjectsResponse{Entries: entries})
+		return
+	}
+
+	// return a meaningful status code depending on whether the parent exists or not
+	key := strings.TrimPrefix(path, "/")
+	_, err = b.ms.Object(jc.Request.Context(), key)
+	if errors.Is(err, api.ErrObjectNotFound) {
+		jc.Error(err, http.StatusNotFound)
+		return
+	} else if err != nil {
+		jc.Error(err, http.StatusInternalServerError)
+		return
+	} else {
+		jc.Error(fmt.Errorf("object at '%v' is not a directory", key), http.StatusBadRequest)
```

Comment:
Since we allow for creating a file and folder with the same name at the same time I think returning anything but a 404 here is confusing. Because on one hand we don't allow for querying a folder that has the same name as a file but we do allow creating a folder like that.

Reply:
If we return a 404 here then fetching the parent object becomes moot? Becomes we return both a 404 if the parent is found yes or no.

Reply:
Ok updated to return a 404 if no entries are found _and_ if no prefix was given. I think as soon as you specify a prefix you would expect an empty list if there's no matches rather than a 404, does that make sense?

Reply:
If we return a 404 when you query a folder that doesn't exist with a prefix, the caller can use the 404 as a hint that the folder doesn't exist instead of assuming that they have chosen a prefix which led to 0 results.
Personally if I get 0 results but no error, I'd expect that by removing the prefix I'd get all results instead of a sudden 404.

Similar to how the Unix command `find` behaves. If you run `find ./foo` and the dir doesn't exist you get an error regardless of the additional filters you provide.